src/DataSets.jl

```diff
@@ -188,21 +201,86 @@ To open a directory as a browsable tree object,
 open(BlobTree, dataset("a_tree_example"))
 ```
 """
-function dataset(proj::AbstractDataProject, name::AbstractString)
-    # Non-fancy search... for now :)
-    # In the future, we can consider parsing `name` into a dataset prefix and a
-    # data selector / resource section. Eg a path for BlobTree which gives us a
-    # SubDataSet
-    #
-    # The URN RFC8141 has some good design inspiration here, in particular the
-    # distinction between r-component and q-component seems relevant:
-    # * Some parameters may need to be passed to the "resolver" (ie, the data
-    #   storage backend)
-    # * Some parameters may need to be passed to the dataset itself (eg, a
-    #   relative path within the dataset)
+function dataset(proj::AbstractDataProject, spec::AbstractString)
+    namestr, querystr, fragmentstr = _split_dataspec(spec)
+
+    if isnothing(namestr)
+        throw(ArgumentError("Invalid dataset specification: $spec"))
+    end
+
+    dataset = proj[namestr]
+
+    if isnothing(querystr) && isnothing(fragmentstr)
+        return dataset
+    end
+
+    # Enhance dataset with "parameters" holding URL-like fragment & query
+    # params
     #
-    # See https://datatracker.ietf.org/doc/html/rfc8141#page-12
-    return proj[name]
+    # We need to take care here to add copy()s to avoid modifying the original
+    # dataset configuration.
+    conf = copy(dataset.conf)
+    params = haskey(conf, "parameters") ? copy(conf["parameters"]) : Dict()
+    conf["parameters"] = params
+    if !isnothing(querystr)
+        query = Dict(split.(split(querystr, '&'), '='))
+        params["query"] = haskey(params, "query") ?
+            merge(query, params["query"]) : query
+    end
+    if !isnothing(fragmentstr)
+        params["fragment"] = fragmentstr
+    end
+
+    return DataSet(conf)
+end
+
+
+# Percent-decode a string according to the URI escaping rules.
+# Vendored from URIs.jl for now to avoid depending on that entire package for
+# this one function.
+function _unescapeuri(str)
+    occursin("%", str) || return str
+    out = IOBuffer()
+    i = 1
+    io = IOBuffer(str)
+    while !eof(io)
+        c = read(io, Char)
+        if c == '%'
+            c1 = read(io, Char)
+            c = read(io, Char)
+            write(out, parse(UInt8, string(c1, c); base=16))
+        else
+            write(out, c)
+        end
+    end
+    return String(take!(out))
+end
+
+function _split_dataspec(spec::AbstractString)
+    # Parse as a suffix of URI syntax
+    # name/of/dataset?param1=value1&param2=value2#fragment
+    m = match(r"
+        ^
+        ((?:[[:alpha:]][[:alnum:]_]*/?)+)  # name     - a/b/c
```

Comment:
Do you want to allow leading slashes?

Reply:
I don't think so, at least not at the moment.

I think `DataSet` names seem more like relative paths, relative to the data project which contains them. If we had data URLs with an authority section, they'd be relative to that authority.